python/cudf/cudf/_libxx/parquet.pyx

```diff
@@ -35,6 +40,40 @@ from cudf._libxx.io.utils cimport (
 cimport cudf._libxx.cpp.types as cudf_types
 cimport cudf._libxx.cpp.io.types as cudf_io_types
 
+cdef class BufferArrayFromVector:
+    cdef unsigned length
+    cdef unique_ptr[vector[uint8_t]] in_vec
+
+    # these two things declare part of the buffer interface
+    cdef Py_ssize_t shape[1]
+    cdef Py_ssize_t strides[1]
+
+    cdef set_ptr(self, unique_ptr[vector[uint8_t]] in_vec):
```

Comment:
Why not make this the `BufferArrayFromVector` constructor, like this? That way it's not possible to replace the underlying `vector<uint8_t>` ptr on existing instances.

```suggestion
    cdef __cinit__(self, unique_ptr[vector[uint8_t]] in_vec):
```

Reply:
I tried this and it didn't work for some reason - let me check exactly why and get back to you here.

Reply:
Right - I abandoned this because I kept running into the following issue when trying to `cdef` the `__cinit__` as opposed to `def`:
```
Special methods must be declared with 'def', not 'cdef'
```
Whereas using `def`, I ran into

```
Cannot convert Python object argument to type 'unique_ptr[vector[uint8_t]]
```
The above was the workaround I ended up using, although I'd be happy to learn that I'm missing something trivial to make it work.

Reply:
Oh I forgot about this. Checking the cython docs, there's this note where they recommend initializing from a static `from_ptr()` method:

> Note All constructor arguments will be passed as Python objects. This implies that non-convertible C types such as pointers or C++ objects cannot be passed into the constructor from Cython code. If this is needed, use a factory function instead that handles the object initialisation. It often helps to directly call __new__() in this function to bypass the call to the __init__() constructor.
>
> See [Instantiation from existing C/C++ pointers](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#existing-pointers-instantiation) for an example.

Reply:
I'll try to resolve this comment and hopefully I can make a new multiline suggestion.